eclair-core/src/main/scala/fr/acinq/eclair/channel/Channel.scala

```diff
@@ -869,84 +872,88 @@ class Channel(val nodeParams: NodeParams, val wallet: EclairWallet, remoteNodeId
-        handleLocalError(InvalidFinalScript(d.channelId), d, Some(remoteShutdown))
-      } else if (Commitments.remoteHasUnsignedOutgoingHtlcs(d.commitments)) {
-        handleLocalError(CannotCloseWithUnsignedOutgoingHtlcs(d.channelId), d, Some(remoteShutdown))
-      } else if (Commitments.remoteHasUnsignedOutgoingUpdateFee(d.commitments)) {
-        handleLocalError(CannotCloseWithUnsignedOutgoingUpdateFee(d.channelId), d, Some(remoteShutdown))
-      } else if (Commitments.localHasUnsignedOutgoingHtlcs(d.commitments)) { // do we have unsigned outgoing htlcs?
-        require(d.localShutdown.isEmpty, "can't have pending unsigned outgoing htlcs after having sent Shutdown")
-        // are we in the middle of a signature?
-        d.commitments.remoteNextCommitInfo match {
-          case Left(waitForRevocation) =>
-            // yes, let's just schedule a new signature ASAP, which will include all pending unsigned changes
-            val commitments1 = d.commitments.copy(remoteNextCommitInfo = Left(waitForRevocation.copy(reSignAsap = true)))
-            // in the meantime we won't send new changes
-            stay using d.copy(commitments = commitments1, remoteShutdown = Some(remoteShutdown))
-          case Right(_) =>
-            // no, let's sign right away
-            self ! CMD_SIGN()
-            // in the meantime we won't send new changes
-            stay using d.copy(remoteShutdown = Some(remoteShutdown))
-        }
-      } else {
-        // so we don't have any unsigned outgoing changes
-        val (localShutdown, sendList) = d.localShutdown match {
-          case Some(localShutdown) =>
-            (localShutdown, Nil)
-          case None =>
-            val localShutdown = Shutdown(d.channelId, d.commitments.localParams.defaultFinalScriptPubKey)
-            // we need to send our shutdown if we didn't previously
-            (localShutdown, localShutdown :: Nil)
-        }
-        // are there pending signed changes on either side? we need to have received their last revocation!
-        if (d.commitments.hasNoPendingHtlcsOrFeeUpdate) {
-          // there are no pending signed changes, let's go directly to NEGOTIATING
-          if (d.commitments.localParams.isFunder) {
-            // we are funder, need to initiate the negotiation by sending the first closing_signed
-            val (closingTx, closingSigned) = Closing.makeFirstClosingTx(keyManager, d.commitments, localShutdown.scriptPubKey, remoteShutdown.scriptPubKey, nodeParams.onChainFeeConf.feeEstimator, nodeParams.onChainFeeConf.feeTargets)
-            goto(NEGOTIATING) using DATA_NEGOTIATING(d.commitments, localShutdown, remoteShutdown, List(List(ClosingTxProposed(closingTx, closingSigned))), bestUnpublishedClosingTx_opt = None) storing() sending sendList :+ closingSigned
+      d.commitments.getRemoteShutdownScript(remoteScriptPubKey) match {
+        case Left(e) =>
+          log.warning("they sent an invalid closing script")
```

Comment:
We should also send a `Warning` message explaining that now that we have this option ;)
Note that in this case, we should delay the `peer ! Peer.Disconnect(remoteNodeId)` to give a chance to our warning to be sent (see https://github.com/ACINQ/eclair/blob/ca51a2d16883d277da41258e527a4ac7a6894bb6/eclair-core/src/main/scala/fr/acinq/eclair/channel/Channel.scala#L1709).

Reply:
done in https://github.com/ACINQ/eclair/pull/1846/commits/e3174057030eb52428df1f77937f56568db4dfe4